app/modules/softuart.c

```diff
@@ -271,125 +262,83 @@ static void softuart_rx_callback(task_param_t arg)
-	if (suart == NULL) {
-		NODE_DBG("Userdata is nil\n");
-		return 0;
-	}
-	stack++;
 
-	const char *method = luaL_checklstring(L, stack, &name_len);
-	if (method == NULL)
-		return luaL_error(L, "Wrong argument type");
-	stack++;
-
-	if (lua_type(L, stack) == LUA_TNUMBER) {
-		suart->softuart->need_len = (uint16_t)luaL_checkinteger( L, stack );
-		stack++;
-		suart->softuart->end_char = 0;
-		if (suart->softuart->need_len > SOFTUART_MAX_RX_BUFF) {
-			suart->softuart->need_len = 0;
-			return luaL_error(L, "Argument bigger than SoftUART buffer");
-		}
-		suart->softuart->armed = 1;
-	} else if (lua_isstring(L, stack)) {
-		const char *end = luaL_checklstring(L , stack,  &arg_len);
-		stack++;
-		if ( arg_len != 1) {
-			return luaL_error(L, "Wrong end char length");
-		}
-		suart->softuart->end_char = end[0];
-		suart->softuart->need_len = 0;
-		suart->softuart->armed = 1;
+	softuart_t *softuart = (softuart_t*)luaL_checkudata(L, 1, "softuart.port");
+	const char *method = luaL_checklstring(L, 2, &name_len);
+
+	if (lua_isnumber(L, 3)) {
+		luaL_argcheck(L, luaL_checkinteger(L, 3) < SOFTUART_MAX_RX_BUFF,
+				2, "Argument bigger than SoftUART buffer");
+		softuart->end_char = 0;
+		softuart->need_len = (uint16_t) luaL_checkinteger(L, 3);
+		softuart->armed = 1;
+	} else if (lua_isstring(L, 3)) {
+		const char *end = luaL_checklstring(L , 3,  &arg_len);
+		luaL_argcheck(L, arg_len == 1, 3, "Wrong end char length");
+		softuart->end_char = end[0];
+		softuart->need_len = 0;
+		softuart->armed = 1;
 	} else {
 		return luaL_error(L, "Wrong argument type");
 	}
 
-
-	if (lua_type(L, stack) == LUA_TFUNCTION) {
-		lua_pushvalue(L, stack); // Copy to top of the stack
-	} else {
-		lua_pushnil(L);
-	}
-
-    if (name_len == 4 && strcmp(method, "data") == 0) {
-
-    	if(suart->softuart->pin_rx == 0xFF) {
-    		return luaL_error(L, "Rx pin was not declared");
-    	}
-
-    	if (softuart_rx_cb_ref[suart->softuart->pin_rx] != LUA_NOREF) {
-    		luaL_unref(L, LUA_REGISTRYINDEX, softuart_rx_cb_ref[suart->softuart->pin_rx]);
-    		softuart_rx_cb_ref[suart->softuart->pin_rx] = LUA_NOREF;
-    	}
-    	if (! lua_isnil(L, -1)) {
-    		softuart_rx_cb_ref[suart->softuart->pin_rx] = luaL_ref(L, LUA_REGISTRYINDEX);
-    	} else {
-    		lua_pop(L, 1);
-    	}
-    } else {
-    	lua_pop(L, 1);
-    	return luaL_error(L, "Method not supported");
+	luaL_argcheck(L, lua_isfunction(L, 4), -1, "No callback function specified");
+	luaL_argcheck(L, (name_len == 4 && strcmp(method, "data") == 0), 2, "Method not supported");
+    luaL_argcheck(L, softuart->pin_rx != 0xFF, 1, "Rx pin was not declared");
+    lua_settop(L, 4); // Move to the top of the stack
+    if (softuart_rx_cb_ref[softuart->pin_rx] != LUA_NOREF) {
+    	// Remove old callback and add new one
+    	luaL_unref(L, LUA_REGISTRYINDEX, softuart_rx_cb_ref[softuart->pin_rx]);
     }
+    // Register callback
+    softuart_rx_cb_ref[softuart->pin_rx] = luaL_ref(L, LUA_REGISTRYINDEX);
```

Comment:
This `if() { unref } ref` dance can be simplified to @TerryE's new `luaL_reref` function:
```C
luaL_reref(L, LUA_REGISTRYINDEX, &softuart_rx_cb_ref[softuart->pin_rx])
```

Reply:
@nwf :+1: Thanks for looking into it - I added this change and `unref2` macro in the newest commit.

Reply:
@galjonsfigur , I apologize, no leak in the module but in my code, at local timer for some reason does not de-register, a simple `tmr=nil` did the trick. Thanks for helping.